### PR TITLE
[fix #193] fix cdc lose data by add resolved ts interval 

### LINKED ISSUE #193: cdc: Data may be lost during leader transfer
### 1. Describe the bug

In TiKV, when a leader transfer occurs, the old leader may send the last resolved ts, which may be larger than the new leader's first causal timestamp after transfer.

Details as follow(from log):

1. At 20:04:06.105, leader transfer from A to B, now B's causal ts is 435080671018356250.
2. At 20:04:06.106, A as a leader broadcasts resolved ts(435080671018356650).
3. At 20:04:06.116, B becomes leader, appends ts(435080671018356350) for new key.
5. At 20:04:06.132, cdc receives resolved ts(435080671018356650) from A.
6. At 20:04:06.225, cdc start new request to B with start ts(435080671018356650).

So B don't send the key that is with ts(435080671018356350) to cdc.

### 2. Minimal reproduce step (Required)
1. start cdc
2. put 5000000 key/value pairs into src tikv
3. get checksum from src tikv and dst tikv
4. if checksums are same, repeat step2-3. 

### 3. What did you see instead (Required)
1. checksums are not same

### 4. What did you expect to see? (Required)
1. checksums are not always same

### PR DESCRIPTION
**Pull request by zeminzhou**
Signed-off-by: zeminzhou <zhouzemin@pingcap.com>

### What problem does this PR solve?

TiKV-CDC maybe lose data when region leader tranfer happend

Issue Number: close #193

Problem Description: **TBD**

### What is changed and how does it work?

Add a safe interval for resolved ts

### Code changes

- Has exported function/method change

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test
- Manual test (add detailed scripts or steps below)

### CODE REVIEW

File: cdc/pkg/config/kvclient.go [hunk 1]
```diff
@@ -21,4 +23,6 @@ type KVClientConfig struct {
 	WorkerPoolSize int `toml:"worker-pool-size" json:"worker-pool-size"`
 	// region incremental scan limit for one table in a single store
 	RegionScanLimit int `toml:"region-scan-limit" json:"region-scan-limit"`
+	// the safe interval to move forward resolved ts
+	ResolvedTsSafeInterval time.Duration `toml:"reoslvedts-safe-interval" json:"reoslvedts-safe-interval"`
```
Comment (pingyu) [inline cdc/pkg/config/kvclient.go:27]:
```suggestion
	ResolvedTsSafeInterval time.Duration `toml:"reoslved-ts-safe-interval" json:"reoslved-ts-safe-interval"`
```

File: cdc/cdc/kv/region_worker.go [hunk 2]
```diff
@@ -828,3 +836,17 @@ func RunWorkerPool(ctx context.Context) error {
 	})
 	return errg.Wait()
 }
+
+func GetSafeResolvedTs(resolvedTs uint64) uint64 {
+	cfg := config.GetGlobalServerConfig().KVClient
+
+	logicalTs := oracle.ExtractLogical(resolvedTs)
+	time := oracle.GetTimeFromTS(resolvedTs)
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:844]: Suggest to use another name. `time` will shadow the "time" package.

File: cdc/cdc/kv/region_worker.go [hunk 3]
```diff
@@ -828,3 +836,17 @@ func RunWorkerPool(ctx context.Context) error {
...
+
+	safeTime := time.Add(-cfg.ResolvedTsSafeInterval)
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:846]: If `time` is less than `cfg.ResolvedTsSafeInterval`, suggest to just return the `time`.
A `resolved-ts = 0` would be a legal one ?

File: cdc/cdc/kv/region_worker.go [hunk 4]
```diff
@@ -828,3 +836,17 @@ func RunWorkerPool(ctx context.Context) error {
 	})
 	return errg.Wait()
 }
+
+func GetSafeResolvedTs(resolvedTs uint64) uint64 {
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:840]: Add unit test for the new method.

File: cdc/cdc/kv/region_worker.go [hunk 5]
```diff
@@ -730,31 +730,39 @@ func (w *regionWorker) handleResolvedTs(
 		return nil
 	}
 	regionID := state.sri.verID.GetID()
+
+	safeResolvedTs := GetSafeResolvedTs(resolvedTs)
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:734]: Please comment that why we need the safeResolvedTs.

File: cdc/cdc/kv/region_worker.go [hunk 6]
```diff
@@ -730,31 +730,39 @@ func (w *regionWorker) handleResolvedTs(
...
+	if safeResolvedTs == 0 {
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:735]: If we accept a `resolvedTs` less that safe interval, suggest to move the warning into `GetSafeResolvedTs`.

File: cdc/cdc/kv/region_worker.go [hunk 7]
```diff
@@ -730,31 +730,36 @@ func (w *regionWorker) handleResolvedTs(
 		return nil
 	}
 	regionID := state.sri.verID.GetID()
+
+	// In TiKV, hen a leader transfer occurs, the old leader may send the last
+	// resolved ts, which may be larger than the new leader appends key to ts.
+	// So we fallback the resolved ts to a safe interval to make sure it's correct.
+	safeResolvedTs := GetSafeResolvedTs(resolvedTs)
```
Comment (haojinming) [inline cdc/cdc/kv/region_worker.go:737]: Why not just replace `resolvedTs` with `safeResolvedTs`?

File: cdc/pkg/config/config_test_data.go [hunk 8]
```diff
@@ -81,6 +81,7 @@ const (
     "worker-concurrent": 8,
     "worker-pool-size": 0,
     "region-scan-limit": 40
+	"resolvedts-safe-interval": 3000000000,
```
Comment (haojinming) [inline cdc/pkg/config/config_test_data.go:84]: Suggest use time.Duration(3) * time.Second

File: cdc/pkg/config/kvclient.go [hunk 9]
```diff
@@ -21,4 +23,6 @@ type KVClientConfig struct {
 	WorkerPoolSize int `toml:"worker-pool-size" json:"worker-pool-size"`
 	// region incremental scan limit for one table in a single store
 	RegionScanLimit int `toml:"region-scan-limit" json:"region-scan-limit"`
+	// the safe interval to move forward resolved ts
+	ResolvedTsSafeInterval time.Duration `toml:"reoslved-ts-safe-interval" json:"reoslvedts-safe-interval"`
```
Comment (haojinming) [inline cdc/pkg/config/kvclient.go:27]:
```suggestion
	ResolvedTsSafeInterval time.Duration `toml:"reoslved-ts-safe-interval" json:"reoslved-ts-safe-interval"`
```

File: cdc/pkg/config/config_test_data.go (hunk 8 above)
Comment (haojinming) [reply cdc/pkg/config/config_test_data.go:84]:
```suggestion
	"resolved-ts-safe-interval": time.Duration(3) * time.Second,
```

Comment (haojinming) [inline cdc/pkg/config/config_test_data.go:84]:
```suggestion
	"resolved-ts-safe-interval": 3s,
```

Comment (zeminzhou) [reply cdc/pkg/config/config_test_data.go]: 3 * time.Second will convert to 3000000000 after serialization

**Review by haojinming — APPROVED** [review]
LGTM~

File: cdc/pkg/config/config_test_data.go (hunk 8 above)
Comment (haojinming) [reply cdc/pkg/config/config_test_data.go]: Get it.

File: cdc/cdc/kv/region_worker.go [hunk 10]
```diff
@@ -730,6 +730,12 @@ func (w *regionWorker) handleResolvedTs(
 		return nil
 	}
 	regionID := state.sri.verID.GetID()
+
+	// In TiKV, when a leader transfer occurs, the old leader may send the last
+	// resolved ts, which may be larger than the new leader appends key to ts.
+	// So we fallback the resolved ts to a safe interval to make sure it's correct.
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:736]:
```suggestion
	// So we fallback the resolved ts to a safe interval to make sure it's correct.
    // See https://github.com/tikv/migration/issues/193.
    // TODO: fix the issue completely.
```

Also suggest to write a full analysis in #193, to help us discuss with others, and completely fix it later.

File: cdc/cdc/kv/region_worker.go [hunk 11]
```diff
@@ -730,6 +730,12 @@ func (w *regionWorker) handleResolvedTs(
 		return nil
 	}
 	regionID := state.sri.verID.GetID()
+
+	// In TiKV, when a leader transfer occurs, the old leader may send the last
+	// resolved ts, which may be larger than the new leader appends key to ts.
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker.go:735]:
```suggestion
    // resolved ts, which may be larger than the new leader's first causal timestamp after transfer.
```

File: cdc/cdc/kv/region_worker_test.go [hunk 12]
```diff
@@ -135,3 +135,30 @@ func (s *regionWorkerSuite) TestRegionWorkerPoolSize(c *check.C) {
...
+		},
+		{
+			resolvedTs: (1<<18)*3*1000 + 1,
+			expected:   1,
+		},
+		{
+			resolvedTs: (1<<18)*4*1000 + 1,
+			expected:   (1<<18)*1000 + 1,
+		},
+	}
```
Comment (pingyu) [inline cdc/cdc/kv/region_worker_test.go:158]: Suggest to add boundary cases, e.g:
1. `resolvedTs = 0`
2. Physical clock less than `3s`.
3. Physical clock equals to `3s`.

File: cdc/pkg/config/kvclient.go [hunk 13]
```diff
@@ -21,4 +23,6 @@ type KVClientConfig struct {
 	WorkerPoolSize int `toml:"worker-pool-size" json:"worker-pool-size"`
 	// region incremental scan limit for one table in a single store
 	RegionScanLimit int `toml:"region-scan-limit" json:"region-scan-limit"`
+	// the safe interval to move forward resolved ts
```
Comment (pingyu) [inline cdc/pkg/config/kvclient.go:26]: Move "backward" ?

File: cdc/cdc/kv/region_worker.go (hunk 10 above)
Comment (zeminzhou) [reply cdc/cdc/kv/region_worker.go:736]: OK

**Comment by zeminzhou** [comment]
@pingyu PTAL~

**Review by pingyu — APPROVED** [review]
LGTM~

**Pull request merged**